File: app/src/main/java/com/wyang/study/utils/NodeSeeker.java
package com.wyang.study.utils;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by weiyang on 2017/11/27.
 */
public class NodeSeeker {

    public interface Visitor {//访问器

        void visitNode(TreeNode treeNode);
    }

    public interface Filter {//筛选器

        void filter(TreeNode treeNode, Visitor visitor);
    }

    public interface Predicate {//判断节点是否符合条件

        boolean match(TreeNode treeNode);
    }

    public static NodeSeeker newInstance(TreeNode node) {
        return new NodeSeeker(node);
    }

    public static NodeSeeker newInstance(List<TreeNode> nodes) {
        return new NodeSeeker(nodes);
    }

    private NodeSeeker(TreeNode treeNode) {
        currentFilter = new NodeFilter();
        firstFilter = currentFilter;
        doFilter((node, visitor) -> {
            if (treeNode != null) {
                visitor.visitNode(treeNode);
            }
        });
    }

    private NodeSeeker(List<TreeNode> nodes) {
        currentFilter = new NodeFilter();
        firstFilter = currentFilter;
        doFilter((treeNode, visitor) -> {
            for (TreeNode node : nodes) {
                visitor.visitNode(node);
            }
        });
    }

    public NodeSeeker children() {
        return doFilter((node, visitor) -> {
            for (TreeNode treeNode : node.getChildren()) {
                visitor.visitNode(treeNode);
            }
        });
    }

    public NodeSeeker hasAttribute(String key) {
        return matchPredicate(treeNode -> treeNode.attributes.containsKey(key));
    }

    public NodeSeeker attribute(String key, String value) {
        return matchPredicate(treeNode -> TextUtils.equals(treeNode.getAttribute(key), value));
    }

    public NodeSeeker attribute(String key, int value) {
        return matchPredicate(treeNode -> (int) treeNode.getAttribute(key) == value);
    }

    public NodeSeeker descendants() {
        return doFilter((node, visitor) -> {
            Visitor[] temp = {null};
            temp[0] = (tt) -> {
                visitor.visitNode(tt);
                for (TreeNode ttChild : tt.getChildren())
                    temp[0].visitNode(ttChild);
            };
            temp[0].visitNode(node);
        });
    }

    public NodeSeeker matchPredicate(Predicate predicate) {
        return doFilter((treeNode, visitor) -> {
            if (predicate == null || predicate.match(treeNode)) {
                visitor.visitNode(treeNode);
            }
        });
    }

    public NodeSeeker leaf() {
        return descendants().matchPredicate(treeNode -> treeNode.size() == 0);
    }

    public NodeSeeker parent() {
        return doFilter((treeNode, visitor) -> {
            if (treeNode.parent != null) {
                visitor.visitNode(treeNode.parent);
            }
        });
    }

    public NodeSeeker unique() {
        Set<TreeNode> set = new HashSet<>();
        return doFilter((treeNode, visitor) -> {
            if (!set.contains(treeNode)) {
                set.add(treeNode);
                visitor.visitNode(treeNode);
            }
        });
    }

    private void execute(Visitor visitor) {
        currentFilter.nextVisitor = visitor;
        firstFilter.nextVisitor.visitNode(null);
    }

    public TreeNode firstResult() {
        final TreeNode[] treeNode = {null};
        try {
            execute((node) -> {
                treeNode[0] = node;
                throw new BreakException();
            });
        } catch (BreakException ignored) {
        }
        return treeNode[0];
    }

    public List<TreeNode> results() {
        List<TreeNode> list = new ArrayList<>();
        execute(list::add);
        return list;
    }

    private NodeFilter currentFilter;
    private final NodeFilter firstFilter;

    private static class NodeFilter {
        Visitor nextVisitor;
        Filter filter;
    }

    private NodeSeeker doFilter(Filter filter) {
        NodeFilter nodeFilter = new NodeFilter();
        nodeFilter.filter = filter;
        currentFilter.nextVisitor = node -> nodeFilter.filter.filter(node, nodeFilter.nextVisitor);
        currentFilter = nodeFilter;
        return this;
    }

    /**
     * 强制终止异常
     */
    private static class BreakException extends RuntimeException {

    }
}

File: app/src/main/java/com/wyang/study/NodeSeeker.kt
package com.wyang.study

import android.util.Log

typealias Visitor = (node: TreeNode?) -> Unit
typealias Predicate = (node: TreeNode) -> Boolean
typealias Filter = (node: TreeNode, visitor: Visitor) -> Unit

open class NodeSeeker(node: TreeNode?) {
    companion object {
        fun newInstance(node: TreeNode): NodeSeeker {
            return NodeSeeker(node)
        }
    }

    private var currentFilter = NodeFilter()
    private var firstFilter = currentFilter

    init {
        doFilter { _, function ->
            node?.let {
                function.invoke(it)
            }
        }
    }

    open fun children() = doFilter { node, function ->
        node.children.forEach {
            function.invoke(it)
        }
    }

    open fun match(predicate: Predicate) = doFilter { node, function ->
        if (predicate(node)) {
            function.invoke(node)
        }
    }

    fun descendants() = doFilter { node, function ->
        var visitor: Visitor? = null
        visitor = { tt ->
            function.invoke(tt)
            node.children.forEach {
                visitor?.invoke(it)
            }
        }
        visitor.invoke(node)
    }

    private fun doFilter(filter: Filter): NodeSeeker {
        val nodeFilter = NodeFilter()
        nodeFilter.filter = filter
        currentFilter.visitor = {
            nodeFilter.filter(it!!, nodeFilter.visitor)
        }
        currentFilter = nodeFilter
        return this
    }

    private fun execute(visitor: Visitor) {
        currentFilter.visitor = visitor
        firstFilter.visitor(null)
    }

    fun firstResult(): TreeNode {
        val nodes = mutableListOf<TreeNode>()
        execute {
            Log.e("node111111111111111", it.toString())
            it?.let {
                nodes[0] = it
            }
        }
        Log.e("node111111111111111", nodes.toString())

        return nodes[0]
    }

    fun results(): MutableList<TreeNode> {
        val list = mutableListOf<TreeNode>()
        execute {
            it?.let {
                list.add(it)
            }
        }
        return list
    }

    class NodeFilter {
        lateinit var visitor: Visitor
        lateinit var filter: Filter
    }
}